File: elixir/hamming/hamming.exs
defmodule Hamming do
  @doc """
  Returns number of differences between two strands of DNA, known as the Hamming Distance.

  ## Examples

  iex> Hamming.hamming_distance('AAGTCATA', 'TAGCGATC')
  {:ok, 4}
  """
  @spec hamming_distance([char], [char]) :: non_neg_integer
  def hamming_distance(a, a) do {:ok, 0} end
  def hamming_distance(a, b) when length(a) != length(b) do
    {:error, "Lists must be the same length"}
  end
  def hamming_distance(strand1, strand2) do
    count = List.zip([strand1, strand2]) |>
            Enum.map(&Tuple.to_list/1) |>
            Enum.filter(fn(list) ->
              Enum.fetch(list, 0) != Enum.fetch(list, 1)
            end) |>
            length
    {:ok, count}
  end
end

File: go/hamming/hamming.go
// Package hamming provides functions to find the hamming distance between 2 strings
package hamming

import "errors"

const testVersion = 6

// Distance returns the hamming distance between input strings
func Distance(a, b string) (int, error) {
	if len(a) != len(b) {
		return -1, errors.New("Strings must be the same length")
	}
	diffCount := 0

	for i := 0; i < len(a); i++ {
		if a[i] != b[i] {
			diffCount += 1
		}
	}

	return diffCount, nil
}
